flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java

```diff
@@ -105,11 +105,11 @@ public UnionInputGate(IndexedInputGate... inputGates) {
 		inputChannelToInputGateIndex = new int[totalNumberOfInputChannels];
 
 		int currentNumberOfInputChannels = 0;
-		for (final IndexedInputGate inputGate : inputGates) {
-			inputGateChannelIndexOffsets[inputGate.getGateIndex()] = currentNumberOfInputChannels;
+		for (int index = 0; index < inputGates.length; index++) {
```

Comment:
This seems to be going in the opposite direction - we are replacing indexing based on the real IDs with those based on the order?

Reply:
Yes you are right. I reverted the change and went into a different direction. The basic issue is that the field inputGates contains all gates in a dense array. We would need a sparse index/map if we access it completely through gateIndex.
There were 3 options:
 - Always use dense, which results in what you have mentioned.
 - Use both, resulting in duplicate fields (current solution).
 - Always use sparse index, where we need to skip null values. Or use map and live with slightly longer accesses.

Reply:
Ok, let's try out the `Map` approach. In case of performance issues, we can later fallback to other solution, like sparse index.